### ISSUE TITLE
Support offline installation of pip

### ISSUE THREAD

**Issue by arjun024** [paketo-community/pip#111]
### Problem

pip: v0.1.0

When I try to pack build the [`default_app`](https://github.com/paketo-community/pip/tree/v0.1.0/integration/testdata/default_app) with offline [cypthon buildpack](https://github.com/paketo-community/cpython) (v0.3.0) and [build-plan buildpack](https://github.com/paketo-community/build-plan) with network turned off, it fails with the following error:
```
pack build offlinepipimage -b <cpython-cnb> -b <pip-cnb> -b <build-plan-cnb> --network=none
...
[builder]   Executing build process
[builder]     Installing Pip 21.0.1
[builder] failed to configure pip:
[builder] Processing /tmp/pip-source841787572
[builder]   Installing build dependencies: started
[builder]   Installing build dependencies: finished with status 'error'
[builder]   ERROR: Command errored out with exit status 1:
[builder]    command: /layers/paketo-community_cpython/cpython/bin/python /layers/paketo-community_cpython/cpython/lib/python3.8/site-packages/pip install --ignore-installed --n
o-user --prefix /tmp/pip-build-env-pflm9poq/overlay --no-warn-script-location --no-binary :none: --only-binary :none: -i https://pypi.org/simple -- setuptools wheel
[builder]        cwd: None
[builder]   Complete output (7 lines):
[builder]   WARNING: Retrying (Retry(total=4, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connecti
on.HTTPSConnection object at 0x7f9a8fce0730>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution')': /simple/setuptools/
[builder]   WARNING: Retrying (Retry(total=3, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connecti
on.HTTPSConnection object at 0x7f9a8fce0970>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution')': /simple/setuptools/
[builder]   WARNING: Retrying (Retry(total=2, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connecti
on.HTTPSConnection object at 0x7f9a8fce0b20>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution')': /simple/setuptools/
[builder]   WARNING: Retrying (Retry(total=1, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connecti
on.HTTPSConnection object at 0x7f9a8fce0ca0>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution')': /simple/setuptools/
[builder]   WARNING: Retrying (Retry(total=0, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connecti
on.HTTPSConnection object at 0x7f9a8fce0df0>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution')': /simple/setuptools/
[builder]   ERROR: Could not find a version that satisfies the requirement setuptools (from versions: none)
[builder]   ERROR: No matching distribution found for setuptools
[builder]   ----------------------------------------
[builder] ERROR: Command errored out with exit status 1: /layers/paketo-community_cpython/cpython/bin/python /layers/paketo-community_cpython/cpython/lib/python3.8/site-packages
/pip install --ignore-installed --no-user --prefix /tmp/pip-build-env-pflm9poq/overlay --no-warn-script-location --no-binary :none: --only-binary :none: -i https://pypi.org/simple -- setupt
ools wheel Check the logs for full command output.
```

### Proposal

Pip buildpack should support offline installation.
It looks like the dependency of `setuptools` is not provided with this buildpack. It should be provided with either this buildpack or a via a buildpack of its own.

### Expectation
I expect [this offline test](https://github.com/paketo-community/pip/blob/v0.1.0/integration/offline_test.go#L32) to pass.

**Comment by emmjohnson** [comment]
Just hit this while trying to add vendor/offline support for pip-install.

**Comment by arjun024** [comment]
Possible setuptools dependency to evaluate: https://github.com/cloudfoundry/python-buildpack/blob/v1.7.37/manifest.yml#L148-L151

**Comment by thitch97** [comment]
Another potential path forward is to bundle the missing tools into the dependency consumed by the buildpack. There's an example of this in the `pipenv` buildpack, where the [buildpack dependency](https://github.com/paketo-community/pipenv/blob/0dd81399e86d024dd244292ad5d08462329cb5de/buildpack.toml#L19) is actually a bundled set of tools including `pipenv`, as illustrated below:
```
○ → ll
total 35728
drwxr-xr-x   3 pivotal  staff        96 Apr 20 10:02 .
drwxr-xr-x  47 pivotal  staff      1504 Apr 16 15:54 ..
-rw-r--r--   1 pivotal  staff  18291472 Nov  4 08:39 pipenv_2020.11.4_linux_noarch_cflinuxfs3_ca495250.tgz


○ → tar xzf pipenv_2020.11.4_linux_noarch_cflinuxfs3_ca495250.tgz

○ → ll
total 71728
drwxr-xr-x  30 pivotal  staff       960 Apr 20 10:03 .
drwxr-xr-x  47 pivotal  staff      1504 Apr 16 15:54 ..
-rw-r--r--   1 pivotal  staff    765127 Nov  4 08:38 Arpeggio-1.10.1.tar.gz
-rw-r--r--   1 pivotal  staff     13470 Nov  4 08:37 appdirs-1.4.4.tar.gz
-rw-r--r--   1 pivotal  staff    160604 Nov  4 08:38 attrs-20.2.0.tar.gz
-rw-r--r--   1 pivotal  staff    157997 Nov  4 08:36 certifi-2020.6.20.tar.gz
-rw-r--r--   1 pivotal  staff     72498 Nov  4 08:37 configparser-4.0.2.tar.gz
-rw-r--r--   1 pivotal  staff     29670 Nov  4 08:37 contextlib2-0.6.0.post1.tar.gz
-rw-r--r--   1 pivotal  staff    578925 Nov  4 08:37 distlib-0.3.1.zip
-rw-r--r--   1 pivotal  staff     28187 Nov  4 08:36 enum34-1.1.10.tar.gz
-rw-r--r--   1 pivotal  staff      8549 Nov  4 08:37 filelock-3.0.12.tar.gz
-rw-r--r--   1 pivotal  staff     29289 Nov  4 08:37 importlib_metadata-2.0.0.tar.gz
-rw-r--r--   1 pivotal  staff     29000 Nov  4 08:37 importlib_resources-3.3.0.tar.gz
-rw-r--r--   1 pivotal  staff    332271 Nov  4 08:38 invoke-1.4.1.tar.gz
-rw-r--r--   1 pivotal  staff     30287 Nov  4 08:38 parver-0.3.1.tar.gz
-rw-r--r--   1 pivotal  staff     34393 Nov  4 08:37 pathlib2-2.3.5.tar.gz
-rw-r--r--   1 pivotal  staff   1501159 Nov  4 08:36 pip-20.2.4.tar.gz
-rw-r--r--   1 pivotal  staff   5156780 Nov  4 08:36 pipenv-2020.11.4.tar.gz
-rw-r--r--   1 pivotal  staff  18291472 Nov  4 08:39 pipenv_2020.11.4_linux_noarch_cflinuxfs3_ca495250.tgz
-rw-r--r--   1 pivotal  staff     15534 Nov  4 08:37 pytest-runner-5.2.tar.gz
-rw-r--r--   1 pivotal  staff     33311 Nov  4 08:37 scandir-1.10.0.tar.gz
-rw-r--r--   1 pivotal  staff    858770 Nov  4 08:36 setuptools-44.1.1.zip
-rw-r--r--   1 pivotal  staff     48784 Nov  4 08:38 setuptools_scm-4.1.2.tar.gz
-rw-r--r--   1 pivotal  staff      9529 Nov  4 08:37 singledispatch-3.4.0.3.tar.gz
-rw-r--r--   1 pivotal  staff     33917 Nov  4 08:37 six-1.15.0.tar.gz
-rw-r--r--   1 pivotal  staff     78592 Nov  4 08:36 typing-3.7.4.3.tar.gz
-rw-r--r--   1 pivotal  staff   8284462 Nov  4 08:37 virtualenv-20.1.0.tar.gz
-rw-r--r--   1 pivotal  staff      6169 Nov  4 08:36 virtualenv-clone-0.5.4.tar.gz
-rw-r--r--   1 pivotal  staff     59712 Nov  4 08:38 wheel-0.35.1.tar.gz
-rw-r--r--   1 pivotal  staff     13357 Nov  4 08:37 zipp-1.2.0.tar.gz
```

**Comment by emmjohnson** [comment]
We were able to get a test container running by doing the following - 
- built cpython buildpack offline
- ran docker container offline
- copy wheel into docker container
- copy pip into docker container
- installed wheel in container
- installed pip but added these new flags `--no-build-isolation`, `--no-index`, `--find-links=<points to setuptools in cpython layer>`

Right now, there is no easy way to get wheel in pip. We are proposing that we modify core-deps to bundle pip, wheel, and setuptools. We can then use that bundle to do the above steps.

**Comment by ChuckQuinnIV** [comment]
Hi there - hitting this issue with the latest build, just posting in case something has changed.

It looks like the buildpack is now looking in links during configuration, but still tries to reach out to files.pythonhosted.org to download setuptools:

```
Paketo Buildpack for Pip 0.15.4
  Resolving Pip version
    Candidate version sources (in priority order):
      <unknown> -> ""

    Selected Pip version (using <unknown>): 22.2.2

  Executing build process
    Installing Pip 22.2.2
failed to configure pip:
Looking in links: /tmp/pip-source842535390
Processing /tmp/pip-source842535390
  Installing build dependencies: started
  Installing build dependencies: still running...
  Installing build dependencies: finished with status 'error'
  error: subprocess-exited-with-error

  × pip subprocess to install build dependencies did not run successfully.
  │ exit code: 1
  ╰─> [12 lines of output]
      Looking in links: /tmp/pip-source842535390
      Collecting setuptools
        WARNING: Retrying (Retry(total=4, connect=None, read=None, redirect=None, status=None)) after connection broken by 'ReadTimeoutError("HTTPSConnectionPool(host='files.pythonhosted.org', port=443): Read timed out. (read timeout=15)")': /packages/b6/40/353c051f77ee5618adaf1fd96f4f6bae9714ed0a22c7142c01c24eb77fe4/setuptools-65.5.1-py3-none-any.whl
        WARNING: Retrying (Retry(total=3, connect=None, read=None, redirect=None, status=None)) after connection broken by 'ReadTimeoutError("HTTPSConnectionPool(host='files.pythonhosted.org', port=443): Read timed out. (read timeout=15)")': /packages/b6/40/353c051f77ee5618adaf1fd96f4f6bae9714ed0a22c7142c01c24eb77fe4/setuptools-65.5.1-py3-none-any.whl
        WARNING: Retrying (Retry(total=2, connect=None, read=None, redirect=None, status=None)) after connection broken by 'ReadTimeoutError("HTTPSConnectionPool(host='files.pythonhosted.org', port=443): Read timed out. (read timeout=15)")': /packages/b6/40/353c051f77ee5618adaf1fd96f4f6bae9714ed0a22c7142c01c24eb77fe4/setuptools-65.5.1-py3-none-any.whl
        WARNING: Retrying (Retry(total=1, connect=None, read=None, redirect=None, status=None)) after connection broken by 'ReadTimeoutError("HTTPSConnectionPool(host='files.pythonhosted.org', port=443): Read timed out. (read timeout=15)")': /packages/b6/40/353c051f77ee5618adaf1fd96f4f6bae9714ed0a22c7142c01c24eb77fe4/setuptools-65.5.1-py3-none-any.whl
        WARNING: Retrying (Retry(total=0, connect=None, read=None, redirect=None, status=None)) after connection broken by 'ReadTimeoutError("HTTPSConnectionPool(host='files.pythonhosted.org', port=443): Read timed out. (read timeout=15)")': /packages/b6/40/353c051f77ee5618adaf1fd96f4f6bae9714ed0a22c7142c01c24eb77fe4/setuptools-65.5.1-py3-none-any.whl
      ERROR: Could not install packages due to an OSError: HTTPSConnectionPool(host='files.pythonhosted.org', port=443): Max retries exceeded with url: /packages/b6/40/353c051f77ee5618adaf1fd96f4f6bae9714ed0a22c7142c01c24eb77fe4/setuptools-65.5.1-py3-none-any.whl (Caused by ReadTimeoutError("HTTPSConnectionPool(host='files.pythonhosted.org', port=443): Read timed out. (read timeout=15)"))


      [notice] A new release of pip available: 22.2.2 -> 22.3.1
      [notice] To update, run: pip3 install --upgrade pip
      [end of output]

  note: This error originates from a subprocess, and is likely not a problem with pip.
error: subprocess-exited-with-error

× pip subprocess to install build dependencies did not run successfully.
│ exit code: 1
╰─> See above for output.

note: This error originates from a subprocess, and is likely not a problem with pip.

[notice] A new release of pip available: 22.2.2 -> 22.3.1
[notice] To update, run: pip3 install --upgrade pip

error: exit status 1
ERROR: failed to build: exit status 1
```

**Comment by robdimsdale** [comment]
Hi @ChuckQuinnIV - this is unexpected and confusing - especially because we have integration tests that validate that (most of) the buildpacks run in an offline environment. The tests for the pip buildpack passed before we released the last version.

We just released v0.16.0 of the pip buildpack, and we will release a new version of cpython later today. We will release a new version of the Python buildpack (the buildpack that ties all the component buildpacks together) later this week. Perhaps you could try again with the new versions once they're released?

If you're still having issues with the new releases, could you open a new issue?
